shared/rakefiles/xk.rake

```diff
@@ -17,6 +17,23 @@ task :xk, [:cmd, :skip_secret_mgmt, :preserve_stderr] => [:configure, :configure
 
   Rake::Task["set_secrets"].invoke
 
+  sh_filter "sh -c '
```

Comment:
If I understand things correctly and this is permanent, I don't think `xk.rake` is a good place for this code to live, since it is going to be called on every `sh` and `xk` runs. Maybe it should be a separate resource of `istio` module and HPA resources should depend on it?

Reply:
It has to be run every time, because if you delete that resource (`rake destroy_module` for example), GKE will recreate it. So next time you need to import it again, otherwise TF would fail with an error that the resource already exists.

> it should be a separate resource of istio module

It can't, as the TF state of that module has to be modified before that module plan is executed. Once the module is executed it's too late to modify its state.

I don't like it very much here, but given it has to be called every time before `istio` module runs,  I don't have a better place for it - open to suggestions.

Reply:
How about moving this to a new `istio-import` module so it can be then added to dependencies of `istio`?

Reply:
Well, first it seems conceptually very ugly to have one module modify the state of another.

Apart from that - how would you actually do that? Have a local-exec calling terragrunt? So end up with  `terragrunt` executing `terraform` executing `terragrunt` executing `terrafrom`?.  Also I believe terragrunt executes modules in some weird tmp directories, so you would probably have to hardcode/pass paths where the code/istio module is actually located inside the container originally?

Would this work when calling `deploy_module`? (I don't think so - calling `deploy_module[k8s/gpii/flowmanager]` does not run couchdb module & its `finish_cluster` once it has been deployed).

I was thinking about putting this into a module, but for the above reasons I don't think is a good place. Another alternative I was thinking about is to create a separate task for this, but that would actually make it more expensive.

Otherwise, the cost of this once the import happens, is basically one `terraform state pull` which is imho not too bad (also it shows another issue with our current code structure).

Reply:
> (also it shows another issue with our current code structure)

A bit off-topic but how would you solve this problem in the theoretical "pure terraform" version of gpii-infra?

Reply:
> As written, will this code work under common/ where @env/k8s/istio doesn't exist?

Yes it would - I've explained in the comment about checking return code.

> I think there are many tasks that don't care if the tfstate for istio/ is synced with the reality in GKE: sh, any deploy that doesn't touch istio/, any destroy, numerous utility tasks. Am I missing a connection?

Yes you're correct. The response was in the context it's not sufficient to run it only once. To be more precise the code should run every time before task that either directly or indirectly triggers plan or apply of `istio` module.

> Move this code to its own task in xk_configure.rake
> Optionally invoke the new task from :xk, a la skip_secret_mgmt (but I would make it default to off and call it sync_gke_istio_state or something)
> Add the sync_gke_istio_state flag only where it is needed (deploy* mostly)
> ?

Yep, I like this, looks better than the current state.

> A bit off-topic but how would you solve this problem in the theoretical "pure terraform" version of gpii-infra?

It would be easier to implement as there would be only a single entry point for all plan/apply runs (notice we're not even using native rake dependency mechanism, but implementing parametrized logic instead) and also the redundant runs would be avoided.